src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs

```diff
@@ -34,219 +34,142 @@ public KestrelHttpParser(IKestrelTrace log)
-                                {
-                                    RejectRequestLine(span);
-                                }
-                                else
-                                {
-                                    RejectRequest(RequestRejectionReason.UnrecognizedHTTPVersion,
-                                        versionSpan.GetAsciiStringEscaped(32));
-                                }
-                            }
-                        }
-                        break;
-                    case StartLineState.NewLine:
-                        if (data[i] != ByteLF)
+                var queryStart = i;
+                // Query string
+                if (ch == ByteQuestionMark)
+                {
+                    // We have a query string
+                    for (; i < length; i++)
+                    {
+                        ch = data[i];
+                        if (ch == ByteSpace)
                         {
-                            RejectRequestLine(span);
+                            break;
                         }
-                        i++;
+                    }
+                }
 
-                        goto case StartLineState.Complete;
-                    case StartLineState.Complete:
-                        done = true;
-                        break;
+                var targetBuffer = new Span<byte>(data + pathStart, i - pathStart);
+                var query = new Span<byte>(data + queryStart, i - queryStart);
+
+                // Consume space
+                i++;
+
+                // Version
+                var httpVersion = HttpUtilities.GetKnownVersion(data + i, length - i);
+                if (httpVersion == HttpVersion.Unknown)
+                {
+                    RejectUnknownVersion(span, i);
                 }
-            }
 
-            if (!done)
-            {
-                RejectRequestLine(span);
-            }
+                //  After version 8 bytes and cr 1 byte, expect lf
+                if (data[i + 8 + 1] != ByteLF)
+                {
+                    RejectRequestLine(span);
+                }
 
-            var pathBuffer = span.Slice(pathStart, pathEnd - pathStart);
-            var targetBuffer = span.Slice(pathStart, queryEnd - pathStart);
-            var query = span.Slice(queryStart, queryEnd - queryStart);
+                handler.OnStartLine(method, httpVersion, pathEncoded, targetBuffer, pathBuffer, query, customMethod);
 
-            handler.OnStartLine(method, httpVersion, targetBuffer, pathBuffer, query, customMethod);
+                examined = consumed;
+                return true;
+            }
+        }
 
-            consumed = end;
-            examined = consumed;
-            return true;
+        private static bool GetMultiSpan(ref ReadableBuffer buffer, ref Span<byte> span, out ReadCursor end)
+        {
+            if (ReadCursorOperations.Seek(buffer.Start, buffer.End, out end, ByteLF) >= 0)
```

Comment:
Accessing buffer.Start twice vs storing it in a local? That intentional I assume.